src/gui/plugins/tape_measure/TapeMeasure.cc

```diff
@@ -0,0 +1,282 @@
+    public: std::unordered_set<int> placedMarkers;
+
+    /// \brief The current distance between the two points.  This distance
+    /// is updated as the user hovers the end point as well.
+    public: double distance = 0.0;
+
+    /// \brief The namespace that the markers for this plugin are placed in.
+    public: std::string ns = "tape_measure";
+  };
+}
+
+using namespace ignition;
+using namespace gazebo;
+
+/////////////////////////////////////////////////
+TapeMeasure::TapeMeasure()
+  : ignition::gui::Plugin(),
+  dataPtr(std::make_unique<TapeMeasurePrivate>())
+{
+}
+
+/////////////////////////////////////////////////
+TapeMeasure::~TapeMeasure() = default;
+
+/////////////////////////////////////////////////
+void TapeMeasure::LoadConfig(const tinyxml2::XMLElement *)
+{
+  if (this->title.empty())
+    this->title = "Tape measure";
+
+  ignition::gui::App()->findChild<ignition::gui::MainWindow *>
+      ()->installEventFilter(this);
+}
+
+/////////////////////////////////////////////////
+void TapeMeasure::OnMeasure()
+{
+  this->Reset();
+  this->dataPtr->measure = true;
+  QGuiApplication::setOverrideCursor(Qt::CrossCursor);
+}
+
+/////////////////////////////////////////////////
+void TapeMeasure::OnReset()
+{
+  this->Reset();
+}
+
+/////////////////////////////////////////////////
+void TapeMeasure::Reset()
+{
+  this->DeleteMarker(this->dataPtr->startPointId);
+  this->DeleteMarker(this->dataPtr->endPointId);
+  this->DeleteMarker(this->dataPtr->lineId);
+
+  this->dataPtr->currentId = this->dataPtr->startPointId;
+  this->dataPtr->startPoint = ignition::math::Vector3d::Zero;
+  this->dataPtr->endPoint = ignition::math::Vector3d::Zero;
+  this->dataPtr->distance = 0.0;
+  this->dataPtr->measure = false;
+  this->newDistance();
+  QGuiApplication::restoreOverrideCursor();
+}
+
+/////////////////////////////////////////////////
+double TapeMeasure::Distance()
+{
+  return this->dataPtr->distance;
+}
+
+/////////////////////////////////////////////////
+void TapeMeasure::DeleteMarker(int _id)
+{
+  if (this->dataPtr->placedMarkers.find(_id) ==
+      this->dataPtr->placedMarkers.end())
+    return;
+
+  // Delete the previously created marker
+  ignition::msgs::Marker markerMsg;
+  markerMsg.set_ns(this->dataPtr->ns);
+  markerMsg.set_id(_id);
+  markerMsg.set_action(ignition::msgs::Marker::DELETE_MARKER);
+  this->dataPtr->node.Request("/marker", markerMsg);
+  this->dataPtr->placedMarkers.erase(_id);
+}
+
+/////////////////////////////////////////////////
+void TapeMeasure::DrawPoint(int _id,
+    ignition::math::Vector3d &_point, ignition::math::Vector4d &_color)
+{
+  this->DeleteMarker(_id);
+
+  ignition::msgs::Marker markerMsg;
+  markerMsg.set_ns(this->dataPtr->ns);
+  markerMsg.set_id(_id);
+  markerMsg.set_action(ignition::msgs::Marker::ADD_MODIFY);
+  markerMsg.set_type(ignition::msgs::Marker::SPHERE);
+  ignition::msgs::Set(markerMsg.mutable_scale(),
+    ignition::math::Vector3d(0.1, 0.1, 0.1));
```

Comment:
I know this will probably be complicated to achieve with markers, but I'll just make an observation for the future.

The fixed scale means that the points may be too big or too small depending on what's being simulated and the user's zoom level. Maybe they're too big to measure the distance between fingers of a robot hand, but too small when measuring the size of a road. One option would be to scale it according to the zoom level like it's done for the transform controls:

![scale_tape](https://user-images.githubusercontent.com/5751272/99470393-864b9480-28f9-11eb-91e7-6d48090aacf0.gif)

Reply:
Good point, I believe that functionality is coming from [this line](https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo3/src/gui/plugins/scene3d/Scene3D.cc#L1415)?  After thinking about it for a second, I think the effect comes from the world "moving away" and the visual gizmo staying the same size.  @ColeOSRF do you know how this tool is usually  handled in large vs small scenarios and when the user is zooming in and out?